e2e/nomostest/sync.go

```diff
@@ -263,3 +266,58 @@ func statusHasSyncDirAndNoErrors(status v1beta1.Status, sourceType configsync.So
 	}
 	return nil
 }
+
+func resourceGroupHasReconciled(sourceHash string) testpredicates.Predicate {
+	return func(o client.Object) error {
+		if o == nil {
+			return testpredicates.ErrObjectNotFound
+		}
+		rg, ok := o.(*v1alpha1.ResourceGroup)
+		if !ok {
+			return testpredicates.WrongTypeErr(o, &v1alpha1.ResourceGroup{})
+		}
+
+		// Status is disabled for this ResourceGroup, skip all status checks
+		if root.IsStatusDisabled(rg) {
+			return nil
+		}
+
+		if len(rg.Spec.Resources) != len(rg.Status.ResourceStatuses) {
```

Comment:
Do we want to validate that all the resources match in spec & status? Or is that too much here?

Reply:
That's exactly what the code below does.

Reply:
Ah, right. I was kinda confused by the `resourceCount` map. 

It looks like you're validating that there's not more than one entry with the same ID, which seems good. But it's kinda hard to read.

I think there's a way to enforce that in the apiserver with schema validation, but it doesn't look like our CRD has those defined today. I know resources can define unique compound list keys. They added it for server-side apply list merging. I guess that's something we can add later.

What about building a fake `[]ResourceStatus` from the `status.resourceStatuses` and then using semantic.DeepEquals to compare it to the real `status.resources`. Then you could error with a yaml diff that would make it easy to see how they compare?

Reply:
FWIW, this only works if they're sorted, which they might not be. So you might have to DeepCopy `status.resources` and then sort both the actual and expected lists before comparing them.

Reply:
Correct, the lists are not sorted - hence the decision to build a map to determine equality. Also it's not validating that the IDs are unique - this is purely verifying list equality. It's validating that the lists has the same set of IDs, including same number of duplicate IDs if there are duplicates.

Reply:
Alright. I think there's some readability/correctness changed we could make here, but I don't want to hold up our other PRs, and I don't really think it's all that important for general validation we use in every e2e test. 

I wish we could just check the conditions here, and save all the other fields for validating in RG-specific tests, but that's not today's reality.

We can iterate on this later.